browser/playlists/playlists_browsertest.cc

```diff
@@ -0,0 +1,380 @@
+using brave_playlists::PlaylistsService;
+using brave_playlists::PlaylistsServiceFactory;
+using brave_playlists::kPlaylistsIDKey;
+
+namespace {
+
+std::unique_ptr<net::test_server::HttpResponse> HandleRequest(
+    const net::test_server::HttpRequest& request) {
+  std::unique_ptr<net::test_server::BasicHttpResponse> http_response(
+      new net::test_server::BasicHttpResponse());
+  if (request.relative_url == "/valid_thumbnail" ||
+      request.relative_url == "/valid_media_file_1" ||
+      request.relative_url == "/valid_media_file_2") {
+    http_response->set_code(net::HTTP_OK);
+    http_response->set_content_type("image/gif");
+    http_response->set_content("thumbnail");
+  } else {
+    http_response->set_code(net::HTTP_NOT_FOUND);
+  }
+
+  return std::move(http_response);
+}
+
+}  // namespace
+
+class PlaylistsBrowserTest : public InProcessBrowserTest,
+                             public PlaylistsControllerObserver {
+ public:
+  PlaylistsBrowserTest() : weak_factory_(this) {
+    scoped_feature_list_.InitAndEnableFeature(
+        brave_playlists::features::kBravePlaylists);
+  }
+  ~PlaylistsBrowserTest() override {}
+
+  // InProcessBrowserTest overrides:
+  void SetUpOnMainThread() override {
+    InProcessBrowserTest::SetUpOnMainThread();
+
+    host_resolver()->AddRule("*", "127.0.0.1");
+
+    // Set up embedded test server to handle fake responses.
+    https_server_.reset(new net::EmbeddedTestServer(
+        net::test_server::EmbeddedTestServer::TYPE_HTTPS));
+    https_server_->SetSSLConfig(net::EmbeddedTestServer::CERT_OK);
+    https_server_->RegisterRequestHandler(base::BindRepeating(&HandleRequest));
+    ASSERT_TRUE(https_server_->Start());
+
+    GetPlaylistsController()->AddObserver(this);
+    ResetStatus();
+  }
+
+  void TearDownOnMainThread() override {
+    InProcessBrowserTest::TearDownOnMainThread();
+    GetPlaylistsController()->RemoveObserver(this);
+  }
+
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    // HTTPS server only serves a valid cert for localhost, so this is needed
+    // to load pages from other hosts without an error
+    command_line->AppendSwitch(switches::kIgnoreCertificateErrors);
+  }
+
+  // PlaylistsControllerObserver overrides:
+  void OnPlaylistsInitialized(bool initialized) override { run_loop()->Quit(); }
+
+  void OnPlaylistsChanged(const PlaylistsChangeParams& params) override {
+    on_playlists_changed_called_count_++;
+    change_params_ = params;
+    called_change_types_.insert(change_params_.change_type);
+
+    if (change_params_.change_type ==
+        PlaylistsChangeParams::ChangeType::CHANGE_TYPE_ADDED) {
+      lastly_added_playlist_id_ = change_params_.playlist_id;
+    }
+
+    if (on_playlists_changed_called_count_ ==
+        on_playlists_changed_called_target_count_) {
+      run_loop()->Quit();
+    }
+  }
+
+  void OnPlaylistsDownloadRequested(const std::string& url) override {}
+
+  PlaylistsController* GetPlaylistsController() {
```

Comment:
why does the service need a controller? It seems like an unnecessary extra level of abstraction and the PlaylistsService does almost nothing